// fen-chess-board/lib/chess-utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getFileRank = getFileRank;
exports.emptyBoard = emptyBoard;
exports.files = exports.ranks = void 0;

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var ranks = {
  1: 7,
  2: 6,
  3: 5,
  4: 4,
  5: 3,
  6: 2,
  7: 1,
  8: 0
};
exports.ranks = ranks;
var files = {
  a: 0,
  b: 1,
  c: 2,
  d: 3,
  e: 4,
  f: 5,
  g: 6,
  h: 7
};
/**
 * Returns indices for a cell
 * (can be used to access board arrays)
 *
 * Example:
 *   getFileRank("a2") => [0, 6]
 *
 *   2 = 6 because arrays usally are displayed with 0,0 in the upper
 *   left corner
 *
 * @param {string} square - Eg: "a2"
 */

exports.files = files;

function getFileRank(square) {
  var _square = _slicedToArray(square, 2),
      file = _square[0],
      rank = _square[1];

  return [files[file], ranks[rank]];
}

function emptyBoard() {
  var board = [];

  for (var i = 0; i < 8; i++) {
    board[i] = [];
  }

  return board;
}
